File: whiteboard-app/src/main/java/dev/oscarrojas/whiteboard/ws/protocol/AppMessageBinaryEncoder.java
package dev.oscarrojas.whiteboard.ws.protocol;

import dev.oscarrojas.whiteboard.messaging.AppMessage;
import org.springframework.stereotype.Component;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.time.Instant;

@Component
public class AppMessageBinaryEncoder {

    private static final int EPOCH_MILLISECOND_BYTES = 8;
    private static final int CHANNEL_HEADER_BYTES = 1;
    private static final int ACTION_HEADER_BYTES = 1;
    private static final int PAYLOAD_HEADER_BYTES = 4;

    public byte[] encode(AppMessage message) {
        String channel = message.getChannel();
        String action = message.getAction();
        byte[] payload = message.getPayload();

        byte[] channelBytes = channel.getBytes(StandardCharsets.UTF_8);
        byte[] actionBytes = action.getBytes(StandardCharsets.UTF_8);
        byte[] bytes = new byte[
            EPOCH_MILLISECOND_BYTES
                + CHANNEL_HEADER_BYTES
                + channelBytes.length
                + ACTION_HEADER_BYTES
                + actionBytes.length
                + PAYLOAD_HEADER_BYTES
                + payload.length
            ];
        ByteBuffer buffer = ByteBuffer.wrap(bytes);

        // write timestamp (epoch milliseconds)
        buffer.putLong(message.getTimestamp().toEpochMilli());

        // write channel header
        buffer.put((byte) channelBytes.length);

        // write channel bytes
        buffer.put(channelBytes);

        // write action header (2 bytes)
        buffer.put((byte) actionBytes.length);

        // write action bytes
        buffer.put(actionBytes);

        // write payload header
        buffer.putInt(payload.length);

        // write payload bytes
        buffer.put(payload);

        return buffer.array();
    }

    public AppMessage decode(byte[] bytes) throws BinaryDecodingException {
        AppMessage message = new AppMessage();
        ByteBuffer buffer = ByteBuffer.wrap(bytes);

        try {

            int pos = 0;

            // extract timestamp
            message.setTimestamp(Instant.ofEpochMilli(buffer.getLong()));

            // extract channel header
            int channelLength = buffer.get() & 0xFF;

            // extract channel
            byte[] channel = new byte[channelLength];
            for (int i = 0; i < channelLength; i++) {
                channel[i] = buffer.get();
            }
            message.setChannel(new String(channel, StandardCharsets.UTF_8));

            // extract action header
            int actionLength = buffer.get() & 0xFF;

            // extract action
            byte[] action = new byte[actionLength];
            for (int i = 0; i < actionLength; i++) {
                action[i] = buffer.get();
            }
            message.setAction(new String(action, StandardCharsets.UTF_8));

            // extract payload header
            int payloadLength = buffer.getInt();

            // ensure valid input buffer length
            if (EPOCH_MILLISECOND_BYTES
                + CHANNEL_HEADER_BYTES
                + channelLength
                + ACTION_HEADER_BYTES
                + actionLength
                + PAYLOAD_HEADER_BYTES
                + payloadLength
                != bytes.length) {
                throw new BinaryDecodingException(
                    "Unexpected buffer length (headers do not match actual buffer length)");
            }

            // extract payload
            byte[] payload = new byte[payloadLength];
            for (int i = 0; i < payloadLength; i++) {
                payload[i] = buffer.get();
            }
            message.setPayload(payload);

        } catch (IndexOutOfBoundsException e) {
            throw new BinaryDecodingException(
                "Unexpected end of buffer: \n" + e.getMessage());
        }

        return message;
    }

    public AppMessage decode(ByteBuffer bytes) throws BinaryDecodingException {
        return decode(bytes.array());
    }
}

File: whiteboard-client/src/messaging/AppMessageBinaryEncoder.ts
import type { AppMessage } from "./types";

export const EPOCH_MILLISECOND_BYTES = 8;
export const CHANNEL_HEADER_BYTES = 1;
export const ACTION_HEADER_BYTES = 1;
export const PAYLOAD_HEADER_BYTES = 4;

export default class AppMessageBinaryEncoder {
  utf8Encoder = new TextEncoder();
  utf8Decoder = new TextDecoder();

  encode = (message: AppMessage): ArrayBuffer => {
    const channelBytes: Uint8Array = this.utf8Encoder.encode(message.channel);
    const actionBytes: Uint8Array = this.utf8Encoder.encode(message.action);
    const payloadBytes = new Uint8Array(message.payload);

    // create buffer
    const view = new DataView(
      new ArrayBuffer(
        EPOCH_MILLISECOND_BYTES +
          CHANNEL_HEADER_BYTES +
          channelBytes.length +
          ACTION_HEADER_BYTES +
          actionBytes.length +
          PAYLOAD_HEADER_BYTES +
          payloadBytes.length,
      ),
    );

    let bytePos = 0;

    // write timestamp
    view.setBigUint64(bytePos, BigInt(message.timestamp));
    bytePos += 8;

    // write channel header
    view.setUint8(bytePos++, channelBytes.length);

    // write channel bytes
    for (let i = 0; i < channelBytes.length; i++) {
      view.setUint8(bytePos++, channelBytes[i]);
    }

    // write action header
    view.setUint8(bytePos++, actionBytes.length);

    // write action bytes
    for (let i = 0; i < actionBytes.length; i++) {
      view.setUint8(bytePos++, actionBytes[i]);
    }

    // write payload header
    view.setUint32(bytePos, payloadBytes.length);
    bytePos += 4;

    // write payload bytes
    for (let i = 0; i < payloadBytes.length; i++) {
      view.setUint8(bytePos++, payloadBytes[i]);
    }

    return view.buffer;
  };

  decode = (bytes: ArrayBuffer): AppMessage => {
    const view = new DataView(bytes);
    const message = {} as AppMessage;
    let bufPos = 0;

    // extract timestamp
    message.timestamp = Number(view.getBigUint64(bufPos));
    bufPos += 8;

    // extract channel header
    const channelByteLength = view.getUint8(bufPos++);

    // extract channel
    const channelBytes = new Uint8Array(channelByteLength);
    for (let i = 0; i < channelByteLength; i++) {
      channelBytes[i] = view.getUint8(bufPos++);
    }
    message.channel = this.utf8Decoder.decode(channelBytes);

    // extract action header
    const actionByteLength = view.getUint8(bufPos++);

    // extract action
    const actionBytes = new Uint8Array(actionByteLength);
    for (let i = 0; i < actionByteLength; i++) {
      actionBytes[i] = view.getUint8(bufPos++);
    }
    message.action = this.utf8Decoder.decode(actionBytes);

    // extract payload header
    const payloadByteLength = view.getUint32(bufPos);
    bufPos += 4;

    // extract payload
    const payload = new Uint8Array(payloadByteLength);
    for (let i = 0; i < payloadByteLength; i++) {
      payload[i] = view.getUint8(bufPos++);
    }
    message.payload = payload.buffer;

    return message;
  };
}
